Offline storage (#51)
OfficeDev/PnP-OfficeAddins: Samples/Excel.OfflineStorageAddin/README.md

# Using offline storage techniques to cache add-in data

## Summary

This sample demonstrates how you can implement local storage to enable limited functionality for your add-in when a user experiences lost connection.

## Applies to

-  Office on Windows (Word, Excel, PowerPoint)
-  Office on Mac (Word, Excel, PowerPoint)
-  Office on the web (Word, Excel, PowerPoint)

## Prerequisites

Before running this sample, make sure you have npm and Node.js installed on your computer. You can check by running the commands `node -v` and `npm -v` in your terminal. To get this software, go to https://www.npmjs.com/get-npm.

## Solution

Solution | Author(s)
---------|----------
Excel.OfflineStorageAddin | Nancy Wang, Albert Dotson (**Microsoft**)

## Version history

Version  | Date | Comments
---------| -----| --------
1.0  | (update later) | Initial release

## Disclaimer

> **THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.**

Author: Good point, I'll reach out to @davidchesnut about this. Thanks Michelle!

> **THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.**

Reviewer: @mscharlock I'll need to follow up with Vesa to find out the origin of this text. I agree we could make this more friendly so I can return and update this later once I know more. Thanks!

While this add-in gets its data from a local server, implementation of local storage as shown in this sample can be extended to add-ins that get their data from online sources. Furthermore, although this sample runs only in Excel, local storage can be used to offline data across Word, Excel, and PowerPoint.

**Note**: Depending on the type and size of data you wish to offline, you may wish to look into other offline storage options. If you'd like offline capabilities to persist in the file, you may wish to store your add-in's data in [Office Settings](https://docs.microsoft.com/en-us/javascript/api/office/office.settings?view=office-js). Local storage is also limited in that it can only cache up to 5 MB of information. To store larger amounts of data offline, you may wish to consider using [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API); however, as of now, IndexedDB is not supported by all browsers used by Office add-ins, and may cause your add-in to fail on some computers.

## Build and run the sample

> To get this sample, clone or download this repository. In the command line, navigate to the **Excel.OfflineStorageAddin** folder from your root directory.

Reviewer 2: ```suggestion
1. Clone or download this repository. 
2. In the command line, navigate to the **Excel.OfflineStorageAddin** folder from your root directory.

The following code sample shows you how to do these two steps: 
```

```
$ git clone https://github.com/OfficeDev/PnP-OfficeAddins.git
$ cd PnP-OfficeAddins
$ cd Samples
$ cd Excel.OfflineStorageAddin
```

> You can try out this sample by running the following commands in the command line:

Reviewer 2: ```suggestion
3. Try out this sample by running the following commands:
```

```
# this will download the node modules needed to run this add-in
$ npm install

# this will build the add-in 
$ npm run build

# this will start the server to run your add-in on Excel on your desktop
$ npm run start

# this will start the server to run your add-in on Excel on the web
$ npm run start:web
```
## Key parts of this sample

Navigate to ``Excel.OfflineStorageAddin/src/taskpane/taskpane.js`` to find the implementation of local storage described below. 

### Implementing local storage to offline data
The ``Excel.OfflineStorageAddin/src/taskpane/taskpane.js`` file contains the `loadTable()` function, which makes use of local storage to display a table of basketball player stats when a user loses connection.

In the sample code, the `loadTable()` function first checks if the basketball player data was previously cached into local storage, as shown in the code below. If it exists, the data is parsed from JSON into a readable text format before being passed to `createTable()`, a function which creates a table from the given data. 

> ```

Reviewer 2: ```suggestion
```js
```

>         if (localStorage.DraftPlayerData) {

Reviewer 2: You might consider reformatting: not sure how this renders but if the tabs show up you could delete the first one

            var dataObject = JSON.parse(localStorage.DraftPlayerData);
            createTable(dataObject);
        }
```

> If the data was not previously cached, `loadTable()` attempts to access ``sampleData.js``, which is the file containing the data, through an AJAX call. If this attempt is successful, the function passes the data returned from the file to `createTable()` to produce a table. The data is also converted into a JSON object, which is cached into local storage. However, if the function is unable to access the ``sampleData.js`` file, the function returns an error to the console, notifying the add-in. This process is shown in the following code:

Reviewer 2: ```suggestion
If the data wasn't previously cached, `loadTable()` attempts to access the offline data file, *sampleData.js*, through an AJAX call. If this attempt is successful, the function passes the data returned from the file to `createTable()` to produce a table. The data is also converted into a JSON object, which is cached into local storage. However, if the function is unable to access the *sampleData.js* file, the function returns an error to the console. This process is shown in the following code:
```
How will an end user know if an error has occurred? Do you have an error that shows up in the taskpane or something?

```

>         else {

Reviewer 2: Consider reformatting?

            $.ajax({
                dataType: "json",
                url: "sampleData.js",
                success: function (result, status, xhr) {
                    localStorage.DraftPlayerData = JSON.stringify(result);
                    createTable(result);
                },
                error: function (xhr, status, error) {
                    console.log("Player data failed to load with error: " + error);
                }
            });
        }
```

## Security notes

> In the webpack.config.js file, a header is set to  `"Access-Control-Allow-Origin": "*"`. This is only for development purposes. You should lock this header down to only allowed domains in production code.

Reviewer 2: ```suggestion
In the *webpack.config.js* file, a header is set to  `"Access-Control-Allow-Origin": "*"`. This is only for development purposes. In production code, you should indicate the allowed domains and not leave this header open to all domains. 
```

> You will be prompted to install self-signed certificates when you run this sample on your development computer. The certificates are intended only for running and studying this code sample. Do not reuse them in your own code solutions or in production environments.

Reviewer 2: ```suggestion
You'll be prompted to install self-signed certificates when you run this sample and you should accept them while you are developing. Note that these certificates are intended only for running and studying this code sample. Don't reuse them in your own code solutions or in production environments.
```

You can install or uninstall the self-signed certificates by running the following commands in the project folder.

> ```cli

Reviewer 2: ```suggestion
```command&nbsp;line
```

Furthermore, `localStorage` is not secure and should only be used for data that can be made publicly available. To protect add-in data while offline, consider using offline cookies.
As another option, you can store your add-in's data in [`Office.Settings`](https://docs.microsoft.com/en-us/javascript/api/office/office.settings?view=office-js). Using Office.Settings will enable your add-in's offline capabilities to persist within the file (e.g. if you'd like to share your file with others).

## Build and run the sample

1. Clone or download this repository. 
2. In the command line, navigate to the **Excel.OfflineStorageAddin** folder from your root directory.

The following code sample shows you how to do these two steps: 
```command&nbsp;line

$ git clone https://github.com/OfficeDev/PnP-OfficeAddins.git
$ cd PnP-OfficeAddins
$ cd Samples
$ cd Excel.OfflineStorageAddin
```
You can try out this sample by running the following commands:
```command&nbsp;line
# this will download the node modules needed to run this add-in
$ npm install

# this will build the add-in 
$ npm run build

> # this will start the server to run your add-in on Excel on your desktop

Reviewer 3: This comment isn't quite accurate. "npm start" will start the dev-server and also launch Excel by opening an Excel document that loads your add-in.

Author: Good point, I'll update this to more accurately reflect what's going on. Thanks Adam!

    },
```
If the function fails to successfully get the data from the AJAX call, the function first checks if the basketball player data was previously cached into local storage, as shown in the code below. If it exists, the data is parsed from JSON into a readable text format before being passed to `createTable()`, a function which creates a table from the given data. Otherwise, the function returns an error back to the console.

```js
error: function (xhr, status, error) {
    // If the connections fails, checks if "PlayerData" was previously stored in local storage
    if (localStorage.PlayerData) {
        // Retrieves the string saved earlied under the key "PlayerData"
        // and parses it into an object
        let dataObject = JSON.parse(localStorage.PlayerData);

        // Sends the saved data to the table
        createTable(dataObject);
    }
    else {
        console.log("Player data failed to load with error: " + error);
    }
}
```

## Security notes

In the webpack.config.js file, a header is set to  `"Access-Control-Allow-Origin": "*"`. This is only for development purposes. You should lock this header down to only allowed domains in production code.

> 

Reviewer 3: The certificates are not **self-signed**. `office-addin-dev-certs` creates a SSL certificate that is signed by CA certificate and will prompt to install the CA certificate so that the SSL certificate is trusted.

Reviewer: Fixed

You will be prompted to install self-signed certificates when you run this sample on your development computer. The certificates are intended only for running and studying this code sample. Do not reuse them in your own code solutions or in production environments.

> You can install or uninstall the self-signed certificates by running the following commands in the project folder.

Reviewer 3: Again, not **self-signed**.
